pkg/kube/controllers/index_test.go

```diff
@@ -0,0 +1,141 @@
+	ServiceAccount types.NamespacedName
+	Node           string
+}
+
+func TestIndex(t *testing.T) {
+	c := kube.NewFakeClient()
+	informer := c.KubeInformer().Core().V1().Pods().Informer()
+	c.RunAndWait(test.NewStop(t))
+	index := CreateIndex[*corev1.Pod, SaNode](informer, func(pod *corev1.Pod) []SaNode {
+		if len(pod.Spec.NodeName) == 0 {
+			return nil
+		}
+		if len(pod.Spec.ServiceAccountName) == 0 {
+			return nil
+		}
+		return []SaNode{{
+			ServiceAccount: types.NamespacedName{
+				Namespace: pod.Namespace,
+				Name:      pod.Spec.ServiceAccountName,
+			},
+			Node: pod.Spec.NodeName,
+		}}
+	})
+	k1 := SaNode{
+		ServiceAccount: types.NamespacedName{
+			Namespace: "ns",
+			Name:      "sa",
+		},
+		Node: "node",
+	}
+	k2 := SaNode{
+		ServiceAccount: types.NamespacedName{
+			Namespace: "ns",
+			Name:      "sa2",
+		},
+		Node: "node",
+	}
+	assert.Equal(t, index.Lookup(k1), nil)
+	assert.Equal(t, index.Lookup(k2), nil)
+	pod1 := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "pod",
+			Namespace: "ns",
+		},
+		Spec: corev1.PodSpec{
+			ServiceAccountName: "sa",
+			NodeName:           "node",
+		},
+	}
+	pod2 := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "pod2",
+			Namespace: "ns",
+		},
+		Spec: corev1.PodSpec{
+			ServiceAccountName: "sa2",
+			NodeName:           "node",
+		},
+	}
+	pod3 := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "pod3",
+			Namespace: "ns",
+		},
+		Spec: corev1.PodSpec{
+			ServiceAccountName: "sa",
+			NodeName:           "node",
+		},
+	}
+
+	assertIndex := func(k SaNode, pods ...*corev1.Pod) {
+		t.Helper()
+		assert.EventuallyEqual(t, func() []*corev1.Pod { return index.Lookup(k) }, pods, retry.Timeout(time.Second*5))
+	}
+
+	// When we create a pod, we should (eventually) see it in the index
+	c.Kube().CoreV1().Pods("ns").Create(context.Background(), pod1, metav1.CreateOptions{})
+	assertIndex(k1, pod1)
+	assertIndex(k2)
+
+	// Create another pod; we ought to find it as well now.
+	c.Kube().CoreV1().Pods("ns").Create(context.Background(), pod2, metav1.CreateOptions{})
+	assertIndex(k1, pod1) // Original one must still persist
+	assertIndex(k2, pod2) // New one should be there, eventually
+
+	// Create another pod with the same SA; we ought to find multiple now.
+	c.Kube().CoreV1().Pods("ns").Create(context.Background(), pod3, metav1.CreateOptions{})
+	assertIndex(k1, pod1, pod3) // Original one must still persist
+	assertIndex(k2, pod2)       // New one should be there, eventually
+
+	pod1Alt := pod1.DeepCopy()
+	// This can't happen in practice with Pod, but Index supports arbitrary types
+	pod1Alt.Spec.ServiceAccountName = "new-sa"
+
+	keyNew := SaNode{
+		ServiceAccount: types.NamespacedName{
+			Namespace: "ns",
+			Name:      "new-sa",
+		},
+		Node: "node",
+	}
+	c.Kube().CoreV1().Pods("ns").Update(context.Background(), pod1Alt, metav1.UpdateOptions{})
+	assertIndex(k1, pod3)        // Pod should be dropped from the index
```

Comment:
Can you also assert index obj size so that we can ensure it gets deleted from there as well?